File: modules/expo-encryption/ios/ExpoEncryptionModule.swift
import Foundation
import CryptoKit
import ExpoModulesCore

public class ExpoEncryptionModule: Module {
  
  // Function to perform HMAC-SHA256 hash calculation
  public func hmacSHA256(key: String, message: String) -> String {
    // Convert the key and message to Data
    let keyData = Data(key.utf8) // Key data
    let messageData = Data(message.utf8) // Message data
    
    // Create a SymmetricKey from the key
    let symmetricKey = SymmetricKey(data: keyData)
    
    // Perform the HMAC-SHA256 operation using CryptoKit
    let signature = HMAC<SHA256>.authenticationCode(for: messageData, using: symmetricKey)
    
    // Debugging step: print the result as a hexadecimal string
    let hexString = signature.map { String(format: "%02hhx", $0) }.joined()
    print("Generated HMAC-SHA256 hash: \(hexString)")  // Debugging
    
    return hexString  // Return the result as a hex string
  }

  // Define module functions in Expo
  public func definition() -> ModuleDefinition {
    Name("ExpoEncryption")

    // Define the HMAC-SHA256 function for JavaScript to call
    AsyncFunction("hmacSHA256") { (key: String, message: String) -> String in
      return self.hmacSHA256(key: key, message: message)
    }
  }
}

File: modules/expo-encryption/android/src/main/java/expo/modules/encryption/ExpoEncryptionModule.kt
package expo.modules.encryption

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class ExpoEncryptionModule : Module() {
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module
    Name("ExpoEncryption")

    // Defines the HMAC-SHA256 function
    AsyncFunction("hmacSHA256") { key: String, message: String ->
      // Convert the key and message to bytes
      val keySpec = SecretKeySpec(key.toByteArray(), "HmacSHA256")
      val mac = Mac.getInstance("HmacSHA256")
      mac.init(keySpec)

      // Perform the HMAC-SHA256 hashing
      val hashBytes = mac.doFinal(message.toByteArray())

      // Convert hashBytes to a hexadecimal string (matches iOS behavior)
      val hexString = hashBytes.joinToString("") { byte ->
        "%02x".format(byte)
      }

      hexString // Return the hash as a hexadecimal string
    }
  }
}
